File: frontend/src/services/infrastructure.ts
import { api } from "@/lib/api";
import type { CustomFieldDef } from "./clusterRequests";

export interface Cluster {
  id: string;
  name: string;
  description: string;
  descriptionPt: string;
  cpuUsage: number;
  gpuUsage: number;
  memoryUsage: number;
  storageUsage: number;
  status: "online" | "maintenance";
  custom_fields: CustomFieldDef[];
}

export type ClusterInput = Omit<Cluster, "id">;

export interface InfrastructureData {
  clusters: Cluster[];
  resources: string[];
}

export const infrastructureService = {
  get: () => api.get<InfrastructureData>("/api/v1/infrastructure"),
  getCluster: (id: string) => api.get<Cluster>(`/api/v1/infrastructure/${id}`),
  createCluster: (data: ClusterInput) => api.post<Cluster>("/api/v1/infrastructure", data),
  updateCluster: (id: string, data: ClusterInput) => api.put<Cluster>(`/api/v1/infrastructure/${id}`, data),
  removeCluster: (id: string) => api.delete<void>(`/api/v1/infrastructure/${id}`),
};

File: backend/models/infrastructure.py
from typing import Optional
from pydantic import BaseModel


class CustomFieldDef(BaseModel):
    name: str                # internal key
    label: str               # display label (EN)
    labelPt: str             # display label (PT)
    type: str = "text"       # text | number | select | checkbox | date
    options: list[str] = []  # for select type
    required: bool = False


class ClusterBase(BaseModel):
    name: str
    description: str
    descriptionPt: str
    cpuUsage: int = 0
    gpuUsage: int = 0
    memoryUsage: int = 0
    storageUsage: int = 0
    status: str = "online"
    custom_fields: list[CustomFieldDef] = []


class ClusterCreate(ClusterBase):
    pass


class ClusterUpdate(ClusterBase):
    pass


class ClusterOut(ClusterBase):
    id: str
